File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    // Approach1: loop the prices and keep updating the min. prices and max. profits.
    // Time: O(n)
    // Space: O(1)

    int maxProfit(vector<int>& prices) {
        int max_profit = 0;
        int min_price = prices[0];

        for (int i = 1; i < prices.size(); i++) {
            int profit = prices[i] - min_price;
            max_profit = max(max_profit, profit);
            min_price = min(min_price, prices[i]);
        }

        return max_profit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
# find the largest right value that is larger than the current value

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0
        
        max_right_val = prices[-1]
        max_profit = 0
        for i in range(len(prices) - 2, -1, -1):
            max_profit = max(max_profit, max_right_val - prices[i])
            max_right_val = max(max_right_val, prices[i])
        
        return max_profit